components/script/layout_wrapper.rs

```diff
@@ -234,12 +242,23 @@ impl<'ln> TNode for ServoLayoutNode<'ln> {
         old_value - 1
     }
 
-    fn borrow_data(&self) -> Option<AtomicRef<PersistentStyleData>> {
-        self.get_style_data().map(|d| d.borrow())
+    fn get_existing_style(&self) -> Option<Arc<ComputedValues>> {
+        self.borrow_data().and_then(|x| x.style.clone())
     }
 
-    fn mutate_data(&self) -> Option<AtomicRefMut<PersistentStyleData>> {
-        self.get_style_data().map(|d| d.borrow_mut())
+    fn set_style(&self, style: Option<Arc<ComputedValues>>) {
+        self.mutate_data().unwrap().style = style;
+    }
+
+    fn take_pseudo_styles(&self) -> PseudoStyles {
+        use std::mem;
+        let mut tmp = PseudoStyles::default();
+        mem::swap(&mut tmp, &mut self.mutate_data().unwrap().per_pseudo);
+        tmp
+    }
+
+    fn set_pseudo_styles(&self, styles: PseudoStyles) {
+        self.mutate_data().unwrap().per_pseudo = styles;
```

Comment:
`debug_assert!` that the current pseudo styles is empty?

I don't know if it'd be more efficient to make this something like an `Option`, then use `Option::take`. It'd avoid the construction of the hashmap on `take_pseudo_styles`.

Also, probably you want to open an issue to make the pseudo styles a vector (or a `SmallVec<[2; T]> / AutoTArray<2, T>` and not a hashmap.